### PR TITLE
Implementing new Discord features: Interaction & Slash commands

### PR DESCRIPTION
**Pull request by quinchs**
Hello, I've started writing support for the new discord interaction features.

I think we should talk about Slash commands here, There are some questions I think we all have:
- How will Slash commands be written?
- Will they take on the same style as writing normal commands?
- Will Discord.Net.Commands handle invoking and managing Slash Commands?

I have wrote in some methods for the RestClient to get, create, edit, and delete global and guild slash commands, I've also created some interfaces to use for Interaction with discord.

### CODE REVIEW

**Comment by quinchs** [comment]
#1716

File: src/Discord.Net.Rest/DiscordRestApiClient.cs [hunk 1]
```diff
@@ -786,6 +785,97 @@ public async Task RemoveGroupRecipientAsync(ulong channelId, ulong userId, Reque
...
         }
 
+        //Interactions
+        public async Task<ApplicationCommand[]> GetGlobalApplicationCommandsAsync(RequestOptions options = null)
+        {
+            try
+            {
+                return await SendAsync<ApplicationCommand[]>("GET", $"applications/{this.CurrentUserId}/commands", options: options).ConfigureAwait(false);
+            }
+            catch (HttpException ex) { return null; }
```
Comment (asmejkal) [inline src/Discord.Net.Rest/DiscordRestApiClient.cs]: why? just throw the exception

File: src/Discord.Net.WebSocket/DiscordSocketClient.cs [hunk 2]
```diff
@@ -1778,7 +1778,33 @@ private async Task ProcessMessageAsync(GatewayOpCode opCode, int? seq, string ty
                                     }
                                 }
                                 break;
+                            case "INTERACTION_CREATE":
+                                {
+                                    await _gatewayLogger.DebugAsync("Received Dispatch (INVITE_DELETE)").ConfigureAwait(false);
```
Comment (asmejkal) [inline src/Discord.Net.WebSocket/DiscordSocketClient.cs]: wrong log

Comment (quinchs) [reply src/Discord.Net.WebSocket/DiscordSocketClient.cs]: Yes I just saw that, I'm going to update it soon

File: src/Discord.Net.Rest/DiscordRestApiClient.cs (hunk 1 above)
Comment (quinchs) [reply src/Discord.Net.Rest/DiscordRestApiClient.cs]: Will do

**Comment by Still34** [comment]
Thanks for the contribution! However, before the changes can be moved forward to any future code review, please discard any non-essential changes in the csprojs and add proper documentation for **all** public members. Thanks!

**Comment by SubZero0** [comment]
I'm aware the latest pushes added an analyzer that complains a lot, but please ignore it and make your classes consistent with the others (like removing the spaces between properties etc). I'll review when this PR gets closer to completion.

**Comment by Still34** [comment]
Temporarily disabled the style enforcement via https://github.com/discord-net/Discord.Net/commit/36de7b25f336e26e954f244ed94f658bf00f66e3

**Comment by quinchs** [comment]
fae3fbe fixes the above mentioned issues and adds comments to everything except SocketInteraction

**Comment by quinchs** [comment]
If some people want to test the new event in the socket client, and let me know if there's any issues. I've tested it a bit and have found no issues.

There are still some stuff todo though:

- [x] Implement some way to create Application commands ([Read here](https://discord.com/developers/docs/interactions/slash-commands#registering-a-command))
- [x] Implement file uploads for Interaction responses

I think the best way we can implement writing slash command is with the default Discord.Net.Commands command service, I'm not too familiar with that project so if someone else could take a look at it that would be great

**Comment by quinchs** [comment]
I'm thinking more about command service stuff, should Slash Commands be written with attributes?
```cs
[SlashCommand("CommandName", "Description", new Option()
{
    "user",                  // option name
    "a user",               // option description
    OptionType.User, // option type
    true,                     // required?
},
new Option() 
{
    "reason",                // option name
    "the reason",          // option description
    OptionType.String, // option type
    true,                       // required?
})]
public async Task TestCommand(SocketGuildUser user, string reason)
{

}
```

There are a lot of parameters for slash commands so its kind of hard to decide how to implement them. The downside to the  attribute method is that it gets messy really fast

**Comment by FiniteReality** [comment]
> ```
> [SlashCommand("CommandName", "Description", new Option()
> {
>     "user",
>     "a user",
>     OptionType.User,
>     true,
> },
> new Option() 
> {
>     "reason",
>     "the reason",
>     OptionType.String,
>     true,
> })]
> public async Task TestCommand(SocketGuildUser user, string reason)
> {
> 
> }
> ```

Won't work. Attribute parameters must be constants, which means no `new`.

**Comment by quinchs** [comment]
> Won't work. Attribute parameters must be constants, which means no `new`.

Just realized that. This is going to be complicated...

Another possibility is to make a class for your slash command, something like this:
```cs
public class MySlashCommand : SlashCommandBase
{
    private DiscordSocketClient client;
    public MySlashCommand(DiscordSocketClient client)
        : base()
    {
        // Init
        this.client = client;

        // Maybe listen for this command with an event?
        this.Executed += HandleExecute;

        // Register the command here with all the command args?
        base.Create("some arguments");
    }

    public async Task HandleExecute(SocketInteraction arg)
    {
        await arg.RespondAsync("Success!");
    }
}
```

**Comment by asmejkal** [comment]
Nice job! Maybe the command service changes could be a different PR? Since that's a bit separate from the implementation of the API on the client, and probably going to be complicated. Not everyone needs the Discord.Net command service to use this.

**Comment by quinchs** [comment]
> Nice job! Maybe the command service changes could be a different PR? Since that's a bit separate from the implementation of the API on the client, and probably going to be complicated. Not everyone needs the Discord.Net command service to use this.

I agree, Im going to make a *mockup* command service for SocketInteractions as a reference. We do still need a public way of fetching, creating, modifying, and deleting Global/guild application commands. I'm going to work on that now, and then I think we're ready for a review!

**Comment by quinchs** [comment]
I'm unable to test the functionality of the new rest methods until tomorrow so you are all free to go let me know what to fix lol

**Comment by quinchs** [comment]
All the rest functionality works now, and the socket stuff also works.

**Comment by Mijago** [comment]
> I'm thinking more about command service stuff, should Slash Commands be written with attributes?
> 
> ```cs
> [SlashCommand("CommandName", "Description", new Option()
> {
>     "user",                  // option name
>     "a user",               // option description
>     OptionType.User, // option type
>     true,                     // required?
> },
> new Option() 
> {
>     "reason",                // option name
>     "the reason",          // option description
>     OptionType.String, // option type
>     true,                       // required?
> })]
> public async Task TestCommand(SocketGuildUser user, string reason)
> {
> 
> }
> ```
> 
> There are a lot of parameters for slash commands so its kind of hard to decide how to implement them. The downside to the attribute method is that it gets messy really fast

I may be a bit late to the party, but couldn't we parse the currently existing parameter attributes and use them for the SlashCommand?

For example, my current Bot looks like this:
```cs
[Command("item")]
[Summary("Search for an item in the manifest.")]
public async Task ShowItems(
    [Remainder]
    [Name("Name")]
    [Summary("Specify the name of the item here. You can use RegExp to give a more specific search term.")]
    string name)
{
    await ShowItemsLang(name, "en");
}
```

And we could simply modify it to something like this:

```cs
[Command("item")]
[SlashCommand("item")] // NEW
[Summary("Search for an item in the manifest.")]
public async Task ShowItems(
    [Remainder]
    [Required] // NEW (if required)
    [Name("Name")]
    [Summary("Specify the name of the item here. You can use RegExp to give a more specific search term.")]
    string name)
{
    await ShowItemsLang(name, "en");
}
```

This way, we would have all the information your example above would also have. The types are already defined, so I doubt there is a need to re-define them.
* Name and description of the command 
* Name and description of the argument
* Type of the argument
* Whether the argument is required or not

**Comment by Cenngo** [comment]
I think creating choices can also be accomplished by attributes since there are only 2 fields that need to be filled. I don't think there will be a lot of people who will add more than a handful of choices per option so i think this should'nt be that messy, relatively speaking.

```csharp
[SlashCommand("rainbow6", "description")]
    public class Rainbow6
    {
        [SubCommandGroup("player", "description")]
        [SubCommand("casual", "description")]
        public async Task GetCasual ( [Description("")] [Required]string username, 
            [Description("")] [Choices("UPlay", "uplay")][Choices("XBox Live", "xbl")][Choices("PlayStation Network", "psn")]string platform)
        {

        }

        [SubCommandGroup("player", "description")]
        [SubCommand("ranked", "description")]
        public void GetRanked ( [Description("")] [Required]string username, 
            [Choices("UPlay", "uplay")][Choices("XBox Live", "xbl")][Choices("PlayStation Network", "psn")] string platform )
        {

        }

        [SubCommandGroup("operator", "description")]
        [SubCommand("stats", "description")]
        public void GetStats ( [Description("")] [Required]string username,
            [Choices("Seasonal", "seasonal")][Choices("Overall", "overall")]string scope, 
            [Choices("Close", 5)][Choices("Long", 10)]int saesons)
        {

        }
    }
```

**Comment by Mijago** [comment]
@Cenngo
I don't think choices should be (only) hard-coded, as sometimes it would be nice to add/remove choices depending on the permissions or other situational things. Or must they be set at the registration of the command? Either way, setting them at launch would also be good.

**Comment by ZargorNET** [comment]
As mentioned before, using Attributes will get really messy. I mean just look at your own code @Cenngo and now imagine more parameters. 
I personally think we should just register commands at startup in one's main method via a builder.
Something like:
```cs
    client.RegisterSlashCommand(
                new SlashCommandBuilder("testCommand")
                .AddParameter("name", "type", "description", ["choices"])
                .AddParameter(...)
                .MakeGlobal()
                .AddSubCommand(new SlashCommandBuilder(...)...)
                .Handler(typeof(TestCommandHandler))
                .Build()
            )
```
Inside the handler, which is a class that inherits from - let's say - ``SlashModuleBase<CommandContext>`` where you define the methods that get called. In order to map them correctly, we could use Attributes above each method to indicate which (sub)command belongs to which method.
I think that this will result in the most readable code.

**Comment by quinchs** [comment]
@ZargorNET @Mijago @Cenngo I'm going to open a new PR for implementing slash commands to the command service, this pull is just implementing them into the client.

I like the idea of a `SlashCommandBuilder` and having attribute based commands, I feel like we should support both [attribute](https://github.com/discord-net/Discord.Net/pull/1717#issuecomment-751297636) and [class](https://github.com/discord-net/Discord.Net/pull/1717#issuecomment-748422246) based. We could have the command service boil those into a base object that represents a slash command. 

The reason why I think we should support both is both are easy to write, understandable, and can be interpreted easily with reflection. With my bots and custom command service I have each command in its own class and its much more manageable to write commands that way. Another + for class based slash commands is the class *is* the context, meaning your context and useful methods lives in `this`.

Attributes are good because it would be very easy to migrate/add slash commands to your existing commands, meaning you don't have to re-write all your commands. The only down side I see with attributes is there's gonna be a lot of them.

Tomorrow I'l open a PR for discussion and testing the command service with slash commands.

**Comment by Cenngo** [comment]
> As mentioned before, using Attributes will get really messy. I mean just look at your own code @Cenngo and now imagine more parameters.
> I personally think we should just register commands at startup in one's main method via a builder.
> Something like:
> 
> ```cs
>     client.RegisterSlashCommand(
>                 new SlashCommandBuilder("testCommand")
>                 .AddParameter("name", "type", "description", ["choices"])
>                 .AddParameter(...)
>                 .MakeGlobal()
>                 .AddSubCommand(new SlashCommandBuilder(...)...)
>                 .Handler(typeof(TestCommandHandler))
>                 .Build()
>             )
> ```
> 
> Inside the handler, which is a class that inherits from - let's say - `SlashModuleBase<CommandContext>` where you define the methods that get called. In order to map them correctly, we could use Attributes above each method to indicate which (sub)command belongs to which method.
> I think that this will result in the most readable code.

I know, with attrubutes it gets messy really quick. My problem with registering command seperatly is once you seperate the command registration with the command logic you end up with two different points of failure. In larger projects I think it would be really easy to mess this up. It would be harder to maintain and once there is any discrepency between the registered code and the logic, your commands wouldnt work. I mean, i wouldnt say i am an expert on "large projects" but i know the frustration of trying to debug the most obvious mistake.

**Comment by popcron** [comment]
> My problem with registering command seperatly is once you seperate the command registration with the command logic you end up with two different points of failure

Can try placing the important requirements of the command into the class itself, this way you keep both the logic and majority of the registration data in one place, leaving the registration process only responsible for injecting new slash commands and the main code clutter free.
```cs
client.RegisterSlashCommand<FlipACoin>();
...
[Global]
class TestCommand : ISlashCommand
{
    string ISlashCommand.BaseName => "testCommand";
    Parameter[] ISlashCommand.Parameters => new Parameter [ ]
    {
        Parameter.Create<T>("name", "description"),
        Parameter.Create<Color>("color", "The color to use when testing this command"),
        Parameter.Create<bool>("delete", "Should the invoking message be deleted?", optional: true)
    };

    async Task ISlashCommand.Run(IMessage context, Parameter[] parameters) { }
}
```
>  It would be harder to maintain and once there is any discrepency between the registered code and the logic, your commands wouldnt work

I dont think its true that they simply wouldn't work, though I imagine people would just write giant adapters to hide the messiness behind a magic wall. But I agree with the discrepancy concern.

**Comment by quinchs** [comment]
Can we move the writing of slash commands over to #1733 as this PR is just for adding the support of slash command, not writing slash command/slash command service

**Comment by amekudeku** [comment]
do you know when slash commands will be support on the main branch or dev?

**Comment by Francoimora** [comment]
Is it at least available on a nightly build ? We are heavily using slash commands at work everyday so it would be a great addition to our workflow.

**Comment by NotOfficer** [comment]
Check out #1733, but in my opinion its not worthy to include in a nightly build yet.

**Comment by asmejkal** [comment]
> Check out #1733, but in my opinion its not worthy to include in a nightly build yet.

Many people use their own commands service and don't need #1733 to use this.

**Comment by quinchs** [comment]
#1733 is no where near stable enough to use in a production environment, I wouldn't use it unless your working on it. However this PR is ready to be used and (I think) currently being tested/reviewed by some of the lib maintainers. So stay tuned as this might me merged soon™

**Comment by Foxite** [comment]
Until such time that this gets merged, I have made a nuget package of this PR so it can be used in its current state.

https://www.myget.org/feed/foxite/package/nuget/Discord.NetFoxite.WebSocket

**Comment by NotOfficer** [comment]
The issue i see in this PR is that the same instance of slashmodules is used forever and dependency injection isnt implemented.
Ive made a [fork](https://github.com/NotOfficer/Discord.Net) of this which fixes some exceptions and tweaks which "fix" what ive mentioned above.
Since its mainly for personal use, there is a breaking change for registering commands (78ada145245ae7a59ec70a32f08b3364977fa9ed) which prevents getting rate limited when you have over 5 commands but removes the ability to keep existing commands.

@quinchs i think you can take some of the stuff i changed over to this PR.

**Comment by webmilio** [comment]
> The issue i see in this PR is that the same instance of slashmodules is used forever and dependency injection isnt implemented.
> Ive made a [fork](https://github.com/NotOfficer/Discord.Net) of this which fixes some exceptions and tweaks which "fix" what ive mentioned above.
> Since its mainly for personal use, there is a breaking change for registering commands ([78ada14](https://github.com/discord-net/Discord.Net/commit/78ada145245ae7a59ec70a32f08b3364977fa9ed)) which prevents getting rate limited when you have over 5 commands but removes the ability to keep existing commands.
> 
> @quinchs i think you can take some of the stuff i changed over to this PR.

You should definitively do a PR to @quinchs repo on my branch or his.

File: src/Discord.Net.Core/Entities/Interactions/ApplicationCommandOptionChoice.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+            get => _name;
+            set
+            {
+                if(value?.Length > 100)
+                    throw new ArgumentException("Name length must be less than or equal to 100");
+                _name = value;
+            }
+        }
+
+        // Note: discord allows strings & ints as values. how should that be handled?
```
Comment (emre1702) [inline src/Discord.Net.Core/Entities/Interactions/ApplicationCommandOptionChoice.cs:30]: How about a class having a constructor for each type?
Would prefer that instead of not knowing about the restriction until I look into my logs and see the exceptions.

Comment (emre1702) [reply src/Discord.Net.Core/Entities/Interactions/ApplicationCommandOptionChoice.cs:30]:
```csharp
public class StringOrInteger 
{
    public object Value { get; private set; }
    
    public StringOrInteger(int value) => Value = value;
    public StringOrInteger(string value) => Value = value;
    
    public static explicit operator StringOrInteger(int value) => new StringOrInteger(value);
    public static explicit operator StringOrInteger(string value) => new StringOrInteger(value);
}
```

**Comment by PolarizedIons** [comment]
Any way this can be helped along? Would love to be able to use this feature

**Comment by quinchs** [comment]
> Any way this can be helped along? Would love to be able to use this feature

You can try to find issues with this PR but as of right now it's ready for review.

**Comment by GuyInGrey** [comment]
Hey! So in handling slash commands, it appears for me the SocketInteraction.Data has a lot of null fields that shouldn't be null. Here's an example of me just serializing the data:
image

Just using this:

```cs
public static async Task EliteCommandInteraction(SocketInteraction i)
{
    await i.FollowupAsync("```js\n" + JsonConvert.SerializeObject(i.Data, Formatting.Indented) + "\n```");
}
```

One of the things I can think of that this might be different from tests is that I'm mixing sub commands and sub-command groups. Discord supports this, but this PR might not. What I mean by that:

```
Command
  Subcommand
  Subcommand
  SubcommandGroup
    Subcommand
    Subcommand
```

I'm available on the Discord API server if you want to message me for more details.

Edit about an hour later:
Even registering a single command (no command groups involved), and no choices, I'm still getting all null data.
image

File: src/Discord.Net.WebSocket/Entities/Interaction/SocketInteractionDataOption.cs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        /// </summary>
+        public IReadOnlyCollection<SocketInteractionDataOption> Options { get; private set; }
+
+        private DiscordSocketClient discord;
+        private ulong guild;
+
+        internal SocketInteractionDataOption() { }
+        internal SocketInteractionDataOption(Model model, DiscordSocketClient discord, ulong guild)
+        {
+            this.Name = Name;
```
Comment (LXGaming) [inline src/Discord.Net.WebSocket/Entities/Interaction/SocketInteractionDataOption.cs]: This line should be `this.Name = model.Name;`

Comment (GuyInGrey) [reply src/Discord.Net.WebSocket/Entities/Interaction/SocketInteractionDataOption.cs]: I can confirm this fixes the problem I posted about

Comment (quinchs) [reply src/Discord.Net.WebSocket/Entities/Interaction/SocketInteractionDataOption.cs]: Whoops yeah that might be an issue, I'l make a comit tomorrow regarding this and some other changes.

**Comment by quinchs** [comment]
> Even registering a single command (no command groups involved), and no choices, I'm still getting all null data.

I agree that this should not be the case, I believe it is fixable and will look into it tomorrow. 

> I'm available on the Discord API server if you want to message me for more details.

I will keep this in mind as I do need a third eye on some code :D

**Comment by quinchs** [comment]
# Updates and TODO
After looking at the changes with slash commands on discord site, this PR is not merge ready anymore. There are some issues to address:

### Invalid follow up message types
When sending follow up messages, the [API route](https://discord.com/developers/docs/interactions/slash-commands#create-followup-message) is based off of websocket and how the current build parses it is to a RestUserMessage. 

The issue with this is that you should be able to [Modify](https://discord.com/developers/docs/interactions/slash-commands#edit-followup-message) and [Delete](https://discord.com/developers/docs/interactions/slash-commands#delete-followup-message) your follow up messages based off of the docs. Currently this is unsupported. **TODO**

### [Batch edit application command permissions](https://discord.com/developers/docs/interactions/slash-commands#batch-edit-application-command-permissions)
Currently this is not supported at all, this is a **TODO**.

### Get [Guild Application](https://discord.com/developers/docs/interactions/slash-commands#get-guild-application-command-permissions) and [Application](https://discord.com/developers/docs/interactions/slash-commands#get-application-command-permissions) Command permissions
This is currently not supported either, this is a **TODO**.

### [Edit](https://discord.com/developers/docs/interactions/slash-commands#edit-original-interaction-response) and [delete](https://discord.com/developers/docs/interactions/slash-commands#delete-original-interaction-response) original response.
This is also currently unsupported with the current build... Add it to the **TODO**

### [Bulk Overwrite Guild Application Command](https://discord.com/developers/docs/interactions/slash-commands#bulk-overwrite-guild-application-commands)
This is a very useful route and not supported in the build. You guessed it! **TODO**

### [Bulk Overwrite Global Application Command](https://discord.com/developers/docs/interactions/slash-commands#bulk-overwrite-global-application-commands)
Same as above but for global commands. TODO

### [`default_permission`](https://discord.com/developers/docs/interactions/slash-commands#create-global-application-command) for creating/modifying/getting Guild/Global Application commands.
They are in the rest entities, there's just no way to current set/get them. **TODO**

### Better error messages for 400's on creating application commands.
Discord was really kind to include an error json structure telling you exactly what you did wrong when creating application commands. *It would be very helpful* if this was included in the exception when creating application commands. **TODO**

## Conclusion
I'm sure I missed a lot of other small changes that need to be implemented but this gets the point across. This PR is still missing a lot of features and I think we should fix them before setting this pull to ready for review. 

**TL;DR:** PR not done but it works.

If you want to help write any code for this PR you are more than welcome to work on any of the **TODO** items and create a PR over on [my fork](https://github.com/quinchs/Discord.Net/tree/Interactions) and I will merge them here. I can't promise I will work on this pr often as I have school and a software job and unfortunately the Discord.net team doesn't pay me enough 🤣. I will pop in every now an then to do some patches and fix any bugs I find while using it, I will also respond to comments.

**Comment by quinchs** [comment]
> The issue i see in this PR is that the same instance of slashmodules is used forever and dependency injection isnt implemented.
> Ive made a [fork](https://github.com/NotOfficer/Discord.Net) of this which fixes some exceptions and tweaks which "fix" what ive mentioned above.
> Since its mainly for personal use, there is a breaking change for registering commands ([78ada14](https://github.com/discord-net/Discord.Net/commit/78ada145245ae7a59ec70a32f08b3364977fa9ed)) which prevents getting rate limited when you have over 5 commands but removes the ability to keep existing commands.
> 
> @quinchs i think you can take some of the stuff i changed over to this PR.

@NotOfficer After going in depth with your branch, it looks like you solved some of the issues I was experiencing. At some point soon I will start merging the two as some stuff has changed with the api

**Comment by Slamerz** [comment]
> > The issue i see in this PR is that the same instance of slashmodules is used forever and dependency injection isnt implemented.
> > Ive made a [fork](https://github.com/NotOfficer/Discord.Net) of this which fixes some exceptions and tweaks which "fix" what ive mentioned above.
> > Since its mainly for personal use, there is a breaking change for registering commands ([78ada14](https://github.com/discord-net/Discord.Net/commit/78ada145245ae7a59ec70a32f08b3364977fa9ed)) which prevents getting rate limited when you have over 5 commands but removes the ability to keep existing commands.
> > @quinchs i think you can take some of the stuff i changed over to this PR.
> 
> @NotOfficer After going in depth with your branch, it looks like you solved some of the issues I was experiencing. At some point soon I will start merging the two as some stuff has changed with the api

Is the dependency injection issue taken care of in this branch?

**Comment by AraHaan** [comment]
> > ```
> > [SlashCommand("CommandName", "Description", new Option()
> > {
> >     "user",
> >     "a user",
> >     OptionType.User,
> >     true,
> > },
> > new Option() 
> > {
> >     "reason",
> >     "the reason",
> >     OptionType.String,
> >     true,
> > })]
> > public async Task TestCommand(SocketGuildUser user, string reason)
> > {
> > 
> > }
> > ```
> 
> Won't work. Attribute parameters must be constants, which means no `new`.

Not entirely true, you can use ``new string[] {}`` in parameters in attributes (even if it's properties inside the attribute as well it will work).

**Comment by GuyInGrey** [comment]
What about this: Instead of using attributes, have the class inherit an interface. That interface has a method you need to implement that will return the slash command, using some SlashCommandBuilder class.

**Comment by AraHaan** [comment]
> What about this: Instead of using attributes, have the class inherit an interface. That interface has a method you need to implement that will return the slash command, using some SlashCommandBuilder class.

Why not have it like how we make commands currently, but have an optional bool value that if set to true makes it a slash command.

And with it optional it would be by default false so it does not break anyone at all either.

**Comment by GuyInGrey** [comment]
> > What about this: Instead of using attributes, have the class inherit an interface. That interface has a method you need to implement that will return the slash command, using some SlashCommandBuilder class.
> 
> Why not have it like how we make commands currently, but have an optional bool value that if set to true makes it a slash command.
> 
> And with it optional it would be by default false so it does not break anyone at all either.

I mean, from what I'm seeing, there's a lot more to slash commands (Sub command groups and such) than can be effectively created within an attribute

**Comment by nitz** [comment]
> > > What about this: Instead of using attributes, have the class inherit an interface. That interface has a method you need to implement that will return the slash command, using some SlashCommandBuilder class.
> 
> > 
> 
> > Why not have it like how we make commands currently, but have an optional bool value that if set to true makes it a slash command.
> 
> > 
> 
> > And with it optional it would be by default false so it does not break anyone at all either.
> 
> 
> 
> I mean, from what I'm seeing, there's a lot more to slash commands (Sub command groups and such) than can be effectively created within an attribute

Could that not be handled by the way sub commands are handled with the `[Group]` attribute now (Command "[Submodules](https://docs.stillu.cc/guides/commands/intro.html#submodules)")?

**Comment by GuyInGrey** [comment]
> > > > What about this: Instead of using attributes, have the class inherit an interface. That interface has a method you need to implement that will return the slash command, using some SlashCommandBuilder class.
> > 
> > 
> > > 
> > 
> > 
> > > Why not have it like how we make commands currently, but have an optional bool value that if set to true makes it a slash command.
> > 
> > 
> > > 
> > 
> > 
> > > And with it optional it would be by default false so it does not break anyone at all either.
> > 
> > 
> > I mean, from what I'm seeing, there's a lot more to slash commands (Sub command groups and such) than can be effectively created within an attribute
> 
> Could that not be handled by the way sub commands are handled with the `[Group]` attribute now (Command "[Submodules](https://docs.stillu.cc/guides/commands/intro.html#submodules)")?

That could work, yeah. One big class represents the whole slash command. Sub commands are methods, and subcommandgroups are sub-classes, with their own methods as commands.
```
SlashCommand1
  SubCommand1
  SubCommandGroup1
    SubCommand3
```


```cs
[SlashCommand("Name", "Description")]
public class SlashCommand1 : SlashCommand
{
  // Option types are taken from the parameter types. No sense in defining them twice.
  [SlashCommand("Name1", "Description")]
  public async Task SubCommand1([SlashOption("Name", "Description")]SocketGuildUser user) { }
  
  public class SubCommandGroup1 {
    [SlashCommand("Name2", "Description")]
    public async Task SubCommand2() { }
  }
}
```

**Comment by AraHaan** [comment]
I think the best option is to use the pre-existing attributes, and use the ``Groups`` attributes like we have now, but every command in a group must be set to be a slash command in the same current attribute that makes them a normal command (just would have to set the optional value to make them a slash command to true).

**Comment by cfowersorthofi** [comment]
Where can I help with this one? I have a need for slash commands and if I can help drive this to acceptance I'd be happy to get involved.

**Comment by christopherfowers** [comment]
With the recent chatter and the PR owners statements (@quinchs) I’m not sure where this thing actually is in terms of getting pushed along to being usable. Where is slash command support currently? Where is this wonderful group of people trying to take it? And how can I help?

**Comment by quinchs** [comment]
> With the recent chatter and the PR owners statements (@quinchs) I’m not sure where this thing actually is in terms of getting pushed along to being usable. Where is slash command support currently? Where is this wonderful group of people trying to take it? And how can I help?

The lib maintainers are reworking dnet, once thats done il update the PR's as they are currently out of date.

Slash commands and message components are being developed in [Discord.NET Labs](https://github.com/Discord-Net-Labs/Discord.Net-Labs). They are stable enough to use there and thats where the PRs to dnet will originate. If you want to help out you can check the labs repo and join our [Discord server](https://discord.com/invite/dvSfUTet3K)

**Comment by quinchs** [comment]
Closing as this is implemented in #1923

**Comment by leocb** [comment]
Awesome work! Is there a documentation or example project we can use to reference these new features?

**Comment by quinchs** [comment]
> Awesome work! Is there a documentation or example project we can use to reference these new features?

The docs are sort of limited but yes, The [readme](https://github.com/Discord-Net-Labs/Discord.Net-Labs#discordnet-labs) of [Discord.Net Labs](https://github.com/Discord-Net-Labs/Discord.Net-Labs) has some documentation of the new features

**Comment by ghost** [comment]
Hi,

When will this be implemented in Discord.NET instead of labs?

**Comment by leocb** [comment]
AFAIK it's already merged but there's no oficial nuget yet, you can download the source and compile it yourself if you want to start using it

**Comment by Francoimora** [comment]
And when will it be available in the official nuget streamline ? We use CI/CD so it's easier for us to setup if we just have to update the nuget.

**Comment by AraHaan** [comment]
Remora.Discord supports slash commands as well, so if you want you could switch to that.

I have also migrated from Discord.NET and overall it simplified my bot to the point 75% of it's code was due to problems with Discord.NET and hacks to try to mitigate them. After the migrations I was able to nuke all of it and it would work the same.

**Comment by Francoimora** [comment]
That's a shame but I think I'll give it a try soon, thanks.

**Comment by quinchs** [comment]
> AFAIK it's already merged but there's no oficial nuget yet, you can download the source and compile it yourself if you want to start using it

Its currently still in a [PR](https://github.com/discord-net/Discord.Net/pull/1923). its a big pr so it will take a bit to get reviewed and refined

**Comment by ghost** [comment]
Alright, lets hope it will be merged soon and by available.

**Comment by djinnet** [comment]
I am really curious for what is the currently status with the PR? 
It looks really close to be done and all the checks has passed green, so what precisely is holding the big PR back?
